docs/drive-mounts.md

```diff
@@ -0,0 +1,115 @@
+
+### User Interface
+
+The API update will look like:
+
+```
+message CreateVMRequest {
+    // existing fields...
+
+    // Replace "FirecrackerDrive RootDrive" with
+    FirecrackerRootDrive RootDrive;
+
+    // Replace "repeated FirecrackerDrive AdditionalDrives" with
+    repeated FirecrackerDriveMount DriveMounts;
+}
+
+message FirecrackerRootDrive {
+    // (Required) HostPath is the path on the host to the filesystem image or device
+    // that will supply the rootfs of the VM.
+    string HostPath;
+
+    // (Optional) If the HostPath points to a drive or image with multiple
+    // partitions, Partuuid specifies which partition will be used to boot
+    // the VM
+    string Partuuid;
+
+    // (Optional) If set to true, IsReadOnly results in the specified HostPath
+    // being opened as read-only by the Firecracker VMM.
+    bool IsReadOnly;
+
+    // (Optional) RateLimiter configuration that will be applied to the
+    // backing-drive for the VM's rootfs
+    FirecrackerRateLimiter RateLimiter;
+}
+
+message FirecrackerDriveMount {
+    // (Required) HostPath is the path on the host to the filesystem image or device
+    // that will be mounted inside the VM.
+    string HostPath;
+
+    // (Required) VMPath is the path inside the VM guest at which the filesystem
+    // image or device will be mounted.
+    string VMPath;
+
+    // (Required) FilesystemType is the filesystem type (i.e. ext4, xfs, etc.), as
+    // used when mounting the filesystem image inside the VM. The VM guest kernel
+    // is expected to have support for this filesystem.
+    string FilesystemType;
+    
+    // (Optional) Options are fstab-style options that the mount will be performed
+    // within the VM (i.e. ["ro", "noatime"]). Defaults to none if not specified. 
+    // If "ro" is specified, the specified HostPath will be also opened as read-only
+    // by the Firecracker VMM.
+    repeated string Options;
+
+    // (Optional) RateLimiter that will be applied to the backing-drive.
+    FirecrackerRateLimiter RateLimiter;
+}
+```
+
+When `DriveMounts` are specified as part of a `CreateVM` request, Firecracker-containerd will internally take care of loading the filesystem image on the host into a Firecracker drive and then mounting the filesystem on that drive inside the VM guest before the `CreateVM` call returns (i.e. as part of the VM+Agent boot process).
+
+The API update is backwards incompatible, but in such a way that upgrading to the new schema is straightforward. It results in a cleaner interface than trying to adapt the previous schema due to:
+
+1. Removing the need to either support or reject rootfs drives that have the new `VMPath`, `FilesystemType` and `Options` field specified
```

Comment:
The root drive can be non-EXT2 and may have options. Isn't it?

https://github.com/firecracker-microvm/firecracker/blob/80cddef3969e97285418c494ef489fa404e9d859/docs/rootfs-and-kernel-setup.md#creating-a-rootfs-image

> For instance, our getting started guide uses an EXT4 FS image with OpenRC as an init system. Note that, whichever file system you choose to use, support for it will have to be compiled into the kernel, so it can be mounted at boot time.

I cannot see EXT4 dependency, while I haven't tested other filesystems.

Reply:
The root drive specifically is configured entirely by the Firecracker VMM and their API does not support passing filesystem type or mount options: https://github.com/firecracker-microvm/firecracker/blob/master/api_server/swagger/firecracker.yaml#L407-L431

The `FirecrackerDriveMount` objects, on the other hand, are mounted in userspace by our firecracker-containerd-specific agent. When agent does the [mount, it needs to provide the filesystem type and options](http://man7.org/linux/man-pages/man2/mount.2.html).

I believe it's *technically* possible for agent to autodetect the filesystem type before doing the mount (I think the command line `mount` utility does this), but I haven't found a pure-golang library for doing so and don't see that introducing that extra complexity would actually provide much benefit as we'd still want to support passing this `FilesystemType` field even if we managed to make it optional.

Reply:
You are right. Firecracker is doing some special Linux-specific stuff for an root drive.

https://github.com/firecracker-microvm/firecracker/blob/0982209eac8a1bb3eb5f79a762a1413aaabc776c/vmm/src/lib.rs#L825